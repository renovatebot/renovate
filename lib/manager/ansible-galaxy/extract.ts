import * as datasourceGalaxy from '../../datasource/galaxy';
import * as datasourceGitTags from '../../datasource/git-tags';
import { logger } from '../../logger';
import { SkipReason } from '../../types';
import type { PackageDependency, PackageFile } from '../types';

const galaxyUrlRegex = /((git\+)?((git|ssh|http(s)?):\/\/)?(.*@)?([\w.-]+)((:\d+)?\/|:))([\w./-]+)(\.git)?/;
const galaxyRoleRegex = /[\w-]+\.[\w-]+/;

function interpretLine(
  lineMatch: RegExpMatchArray,
  lineNumber: number,
  dependency: PackageDependency
): PackageDependency {
  const localDependency: PackageDependency = dependency;
  const key = lineMatch[2];
  const value = lineMatch[3].replace(/["']/g, '');
  switch (key) {
    case 'name': {
      localDependency.managerData.name = value;
      break;
    }
    case 'version': {
      localDependency.managerData.version = value;
      localDependency.currentValue = value;
      localDependency.managerData.lineNumber = lineNumber;
      break;
    }
    case 'scm': {
      localDependency.managerData.scm = value;
      break;
    }
    case 'src': {
      localDependency.managerData.src = value;
      break;
    }
    default: {
      return null;
    }
  }
  return localDependency;
}

function finalize(dependency: PackageDependency): boolean {
  const dep = dependency;
  if (dependency.managerData.version === null) {
    dep.skipReason = SkipReason.NoVersion;
    return false;
  }

  const source: string = dep.managerData.src;
  const sourceMatch: RegExpMatchArray = galaxyUrlRegex.exec(source);
  if (sourceMatch) {
    dep.datasource = datasourceGitTags.id;
    // remove ending `.git` from URLs
    dep.depName = sourceMatch[10].replace(/\.git$/, '');
    // remove leading `git+` from URLs like `git+https://...`
    dep.lookupName = source.replace(/git\+/, '');
  } else if (galaxyRoleRegex.exec(source)) {
    dep.datasource = datasourceGalaxy.id;
    dep.depName = dep.managerData.src;
    dep.lookupName = dep.managerData.src;
  } else if (galaxyRoleRegex.exec(dep.managerData.name)) {
    dep.datasource = datasourceGalaxy.id;
    dep.depName = dep.managerData.name;
    dep.lookupName = dep.managerData.name;
  } else {
    dep.skipReason = SkipReason.NoSourceMatch;
    return false;
  }
  if (dep.managerData.name !== null) {
    dep.depName = dep.managerData.name;
  }

  return true;
}

export default function extractPackageFile(
  content: string
): PackageFile | null {
  logger.trace('ansible-galaxy.extractPackageFile()');
  const newBlockRegEx = new RegExp(/^\s*-\s*((\w+):\s*(.*))$/);
  const blockLineRegEx = new RegExp(/^\s*((\w+):\s*(.*))$/);
  const deps: PackageDependency[] = [];
  const lines = content.split('\n');
  try {
    for (let lineNumber = 0; lineNumber < lines.length; lineNumber += 1) {
      let lineMatch = newBlockRegEx.exec(lines[lineNumber]);
      if (lineMatch) {
        const dep: PackageDependency = {
          managerData: {
            name: null,
            version: null,
            scm: null,
            src: null,
          },
        };
        do {
          const localdep = interpretLine(lineMatch, lineNumber, dep);
          if (localdep == null) {
            break;
          }
          const line = lines[lineNumber + 1];

          if (!line) {
            break;
          }
          lineMatch = blockLineRegEx.exec(line);
          if (lineMatch) {
            lineNumber += 1;
          }
        } while (lineMatch);
        if (finalize(dep)) {
          delete dep.managerData;
          deps.push(dep);
        }
      }
    }

    if (!deps.length) {
      return null;
    }
    return { deps };
  } catch (err) /* istanbul ignore next */ {
    logger.debug({ err }, 'Error extracting ansible-galaxy deps');
    return null;
  }
}
