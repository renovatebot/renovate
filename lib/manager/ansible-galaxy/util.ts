export const newBlockRegEx = /^\s*-\s*((\w+):\s*(.*))$/;
export const blockLineRegEx = /^\s*((\w+):\s*(.*))$/;
export const galaxyDepRegex = /[\w-]+\.[\w-]+/;
export const dependencyRegex = /^dependencies:/;
export const galaxyRegEx =
  /^\s+(?<lookupName>[\w.]+):\s*["'](?<version>.+)["']\s*/;
export const nameMatchRegex =
  /(?<source>((git\+)?(?:(git|ssh|https?):\/\/)?(.*@)?(?<hostname>[\w.-]+)(?:(:\d+)?\/|:))(?<depName>[\w./-]+)(?:\.git)?)(,(?<version>[\w.]*))?/;
