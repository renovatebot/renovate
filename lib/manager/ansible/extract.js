const { getDep } = require('../dockerfile/extract');

module.exports = {
  extractPackageFile,
};

function extractPackageFile(content) {
  logger.debug('ansible.extractPackageFile()');
  let deps = [];
  let lineNumber = 0;
  for (const line of content.split('\n')) {
    const match = line.match(/^\s*image:\s*'?"?([^\s'"]+)'?"?\s*$/);
    if (match) {
      const currentFrom = match[1];
      const dep = getDep(currentFrom);
      logger.debug(
        {
          depName: dep.depName,
          currentValue: dep.currentValue,
          currentDigest: dep.currentDigest,
        },
        'Docker image inside ansible'
      );
      dep.lineNumber = lineNumber;
      dep.versionScheme = 'docker';
      deps.push(dep);
    }
    lineNumber += 1;
  }
  deps = deps.filter(
    dep => !(dep.currentValue && dep.currentValue.includes('${'))
  );
  if (!deps.length) {
    return null;
  }
  return { deps };
}
