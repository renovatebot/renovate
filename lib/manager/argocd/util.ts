export const keyValueExtractionRegex =
  /^\s*(?<key>[^\s]+):\s+"?(?<value>[^"\s]+)"?\s*$/;
// looks for `apiVersion: argoproj.io/
export const fileTestRegex = /\s*apiVersion:\s*argoproj.io\/\s*/;
