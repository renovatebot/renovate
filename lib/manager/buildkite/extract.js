const { isVersion } = require('../../versioning/semver');

module.exports = {
  extractPackageFile,
};

function extractPackageFile(content) {
  const deps = [];
  try {
    const lines = content.split('\n');
    for (let lineNumber = 0; lineNumber < lines.length; lineNumber += 1) {
      const line = lines[lineNumber];
      const plugins = line.match(/^\s*-?\s*plugins:\s*$/);
      if (plugins) {
        logger.trace(`Matched plugins on line ${lineNumber}`);
        const depLine = lines[lineNumber + 1];
        logger.debug(`serviceImageLine: "${depLine}"`);
        const depLineMatch = depLine.match(/^\s+(?:-\s+)?([^#]+)#([^:]+):/);
        if (depLineMatch) {
          logger.trace('depLineMatch');
          lineNumber += 1;
          const [, depName, currentValue] = depLineMatch;
          let skipReason;
          let repo;
          if (depName.startsWith('https://') || depName.startsWith('git@')) {
            logger.debug({ dependency: depName }, 'Skipping git plugin');
            skipReason = 'git-plugin';
          } else if (!isVersion(currentValue)) {
            logger.debug(
              { currentValue },
              'Skipping non-pinned current version'
            );
            skipReason = 'invalid-version';
          } else {
            const splitName = depName.split('/');
            if (splitName.length === 1) {
              repo = `buildkite-plugins/${depName}-buildkite-plugin`;
            } else if (splitName.length === 2) {
              repo = `${depName}-buildkite-plugin`;
            } else {
              logger.warn(
                { dependency: depName },
                'Something is wrong with buildkite plugin name'
              );
              skipReason = 'unknown';
            }
          }
          const dep = {
            lineNumber,
            depName,
            currentValue,
            skipReason,
          };
          if (repo) {
            dep.datasource = 'github';
            dep.lookupName = repo;
          }
          deps.push(dep);
        }
      }
    }
  } catch (err) /* istanbul ignore next */ {
    logger.error(
      { err, message: err.message },
      'Error extracting buildkite plugins'
    );
  }
  if (!deps.length) {
    return null;
  }
  return { deps };
}
