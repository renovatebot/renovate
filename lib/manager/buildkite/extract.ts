import { logger } from '../../logger';
import { isVersion } from '../../versioning/semver';
import { PackageFile, PackageDependency } from '../common';

export function extractPackageFile(content: string): PackageFile | null {
  const deps: PackageDependency[] = [];
  try {
    const lines = content.split('\n');
    let isPluginsSection = false;
    let pluginsIndent = '';
    for (let lineNumber = 1; lineNumber <= lines.length; lineNumber += 1) {
      const lineIdx = lineNumber - 1;
      const line = lines[lineIdx];
      const pluginsSection = line.match(
        /^(?<pluginsIndent>\s*)(-?\s*)plugins:/
      );
      if (pluginsSection) {
        logger.trace(`Matched plugins on line ${lineNumber}`);
        isPluginsSection = true;
        pluginsIndent = pluginsSection.groups.pluginsIndent;
      } else if (isPluginsSection) {
        logger.debug(`serviceImageLine: "${line}"`);
        const { currentIndent } = line.match(/^(?<currentIndent>\s*)/).groups;
        const depLineMatch = line.match(
          /^\s+(?:-\s+)?(?<depName>[^#]+)#(?<currentValue>[^:]+):/
        );
        if (currentIndent.length <= pluginsIndent.length) {
          isPluginsSection = false;
          pluginsIndent = '';
        } else if (depLineMatch) {
          const { depName, currentValue } = depLineMatch.groups;
          logger.trace('depLineMatch');
          let skipReason: string;
          let repo: string;
          if (depName.startsWith('https://') || depName.startsWith('git@')) {
            logger.debug({ dependency: depName }, 'Skipping git plugin');
            skipReason = 'git-plugin';
          } else if (!isVersion(currentValue)) {
            logger.debug(
              { currentValue },
              'Skipping non-pinned current version'
            );
            skipReason = 'invalid-version';
          } else {
            const splitName = depName.split('/');
            if (splitName.length === 1) {
              repo = `buildkite-plugins/${depName}-buildkite-plugin`;
            } else if (splitName.length === 2) {
              repo = `${depName}-buildkite-plugin`;
            } else {
              logger.warn(
                { dependency: depName },
                'Something is wrong with buildkite plugin name'
              );
              skipReason = 'unknown';
            }
          }
          const dep: PackageDependency = {
            managerData: { lineNumber },
            depName,
            currentValue,
            skipReason,
          };
          if (repo) {
            dep.datasource = 'github';
            dep.lookupName = repo;
          }
          deps.push(dep);
        }
      }
    }
  } catch (err) /* istanbul ignore next */ {
    logger.warn({ err }, 'Error extracting buildkite plugins');
  }
  if (!deps.length) {
    return null;
  }
  return { deps };
}
