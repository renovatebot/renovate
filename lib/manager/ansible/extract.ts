import { logger } from '../../logger';
import * as dockerVersioning from '../../versioning/docker';
import { PackageDependency, PackageFile } from '../common';
import { getDep } from '../dockerfile/extract';

export default function extractPackageFile(
  content: string
): PackageFile | null {
  logger.trace('ansible.extractPackageFile()');
  let deps: PackageDependency[] = [];
  const re = /^\s*image:\s*'?"?([^\s'"]+)'?"?\s*$/;
  for (const line of content.split('\n')) {
    const match = re.exec(line);
    if (match) {
      const currentFrom = match[1];
      const dep = getDep(currentFrom);
      logger.debug(
        {
          depName: dep.depName,
          currentValue: dep.currentValue,
          currentDigest: dep.currentDigest,
        },
        'Docker image inside ansible'
      );
      dep.versioning = dockerVersioning.id;
      deps.push(dep);
    }
  }
  deps = deps.filter((dep) => !dep.currentValue?.includes('${'));
  if (!deps.length) {
    return null;
  }
  return { deps };
}
