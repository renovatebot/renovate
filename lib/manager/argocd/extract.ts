import { loadAll } from 'js-yaml';
import * as gitTags from '../../datasource/git-tags';
import { HelmDatasource } from '../../datasource/helm';
import type { ExtractConfig, PackageDependency, PackageFile } from '../types';
import type { ApplicationDefinition } from './types';
import { fileTestRegex } from './util';

function createDependency(
  definition: ApplicationDefinition
): PackageDependency {
  const source = definition.spec?.source;

  if (source == null) {
    return null;
  }

  // a chart variable is defined this is helm declaration
  if (source.chart) {
    return {
      depName: source.chart,
      registryUrls: [source.repoURL],
      currentValue: source.targetRevision,
      datasource: HelmDatasource.id,
    };
  }
  return {
    depName: source.repoURL,
    currentValue: source.targetRevision,
    datasource: gitTags.id,
  };
}

export function extractPackageFile(
  content: string,
  fileName: string,
  config?: ExtractConfig
): PackageFile | null {
  // check for argo reference. API version for the kind attribute is used
  if (fileTestRegex.test(content) === false) {
    return null;
  }

  const definitions: ApplicationDefinition[] = loadAll(content);

  const deps = definitions
    .map((definition) => createDependency(definition))
    .filter(Boolean);

  return deps.length ? { deps } : null;
}
