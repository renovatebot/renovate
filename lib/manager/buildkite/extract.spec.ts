import { getName, loadFixture } from '../../../test/util';
import { extractPackageFile } from './extract';

const pipeline1 = loadFixture('pipeline1.yml');
const pipeline2 = loadFixture('pipeline2.yml');
const pipeline3 = loadFixture('pipeline3.yml');
const pipeline4 = loadFixture('pipeline4.yml');

describe(getName(), () => {
  describe('extractPackageFile()', () => {
    it('returns null for empty', () => {
      expect(extractPackageFile('nothing here')).toBeNull();
    });
    it('extracts simple single plugin', () => {
      const res = extractPackageFile(pipeline1).deps;
      expect(res).toMatchSnapshot();
      expect(res).toHaveLength(1);
    });
    it('extracts multiple plugins in same file', () => {
      const res = extractPackageFile(pipeline2).deps;
      expect(res).toMatchSnapshot();
      expect(res).toHaveLength(2);
    });
    it('adds skipReason', () => {
      const res = extractPackageFile(pipeline3).deps;
      expect(res).toMatchSnapshot();
      expect(res).toHaveLength(3);
    });
    it('extracts arrays of plugins', () => {
      const res = extractPackageFile(pipeline4).deps;
      expect(res).toMatchSnapshot();
      expect(res).toHaveLength(4);
    });
  });
});
