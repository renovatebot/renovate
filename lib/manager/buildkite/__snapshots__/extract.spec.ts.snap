// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/manager/buildkite/extract extractPackageFile() adds skipReason 1`] = `
Array [
  Object {
    "currentValue": "v1.3.2.5",
    "depName": "namespace/docker-compose",
    "managerData": Object {
      "lineNumber": 5,
    },
    "skipReason": "invalid-version",
  },
  Object {
    "currentValue": "v1.3.2",
    "depName": "https://github.com/buildkite/plugin-docker-compose",
    "managerData": Object {
      "lineNumber": 16,
    },
    "skipReason": "git-plugin",
  },
  Object {
    "currentValue": "v1.3.2",
    "depName": "github.com/buildkite/plugin-docker-compose",
    "managerData": Object {
      "lineNumber": 23,
    },
    "skipReason": "unknown",
  },
]
`;

exports[`lib/manager/buildkite/extract extractPackageFile() extracts arrays of plugins 1`] = `
Array [
  Object {
    "currentValue": "v2.0.1",
    "datasource": "github",
    "depName": "docker-login",
    "lookupName": "buildkite-plugins/docker-login-buildkite-plugin",
    "managerData": Object {
      "lineNumber": 3,
    },
    "skipReason": undefined,
  },
  Object {
    "currentValue": "v2.5.1",
    "datasource": "github",
    "depName": "docker-compose",
    "lookupName": "buildkite-plugins/docker-compose-buildkite-plugin",
    "managerData": Object {
      "lineNumber": 5,
    },
    "skipReason": undefined,
  },
  Object {
    "currentValue": "v2.0.1",
    "datasource": "github",
    "depName": "docker-login",
    "lookupName": "buildkite-plugins/docker-login-buildkite-plugin",
    "managerData": Object {
      "lineNumber": 11,
    },
    "skipReason": undefined,
  },
  Object {
    "currentValue": "v2.5.1",
    "datasource": "github",
    "depName": "docker-compose",
    "lookupName": "buildkite-plugins/docker-compose-buildkite-plugin",
    "managerData": Object {
      "lineNumber": 13,
    },
    "skipReason": undefined,
  },
]
`;

exports[`lib/manager/buildkite/extract extractPackageFile() extracts multiple plugins in same file 1`] = `
Array [
  Object {
    "currentValue": "v1.3.2",
    "datasource": "github",
    "depName": "docker-compose",
    "lookupName": "buildkite-plugins/docker-compose-buildkite-plugin",
    "managerData": Object {
      "lineNumber": 5,
    },
    "skipReason": undefined,
  },
  Object {
    "currentValue": "v1.3.2",
    "datasource": "github",
    "depName": "docker-compose",
    "lookupName": "buildkite-plugins/docker-compose-buildkite-plugin",
    "managerData": Object {
      "lineNumber": 16,
    },
    "skipReason": undefined,
  },
]
`;

exports[`lib/manager/buildkite/extract extractPackageFile() extracts simple single plugin 1`] = `
Array [
  Object {
    "currentValue": "v2.0.0",
    "datasource": "github",
    "depName": "abc/detect-clowns",
    "lookupName": "abc/detect-clowns-buildkite-plugin",
    "managerData": Object {
      "lineNumber": 3,
    },
    "skipReason": undefined,
  },
]
`;
